### ISSUE TITLE
app_textures folder would not clean the jpeg when close the tab

### ISSUE THREAD

**Issue by inorinet** [bromite/bromite#2027]
### Preliminary checklist

- [X] I have read the [README](https://github.com/bromite/bromite/blob/master/README.md).
- [X] I have searched the existing issues for my problem. This is a new ticket, NOT a duplicate or related to another open issue.
- [X] I have read the [FAQs](https://github.com/bromite/bromite/blob/master/FAQ.md).
- [X] I have updated Bromite to the latest version. The bug is reproducible on this latest version.

### Can the bug be reproduced with corresponding Chromium version?

No

### Bromite version

Version 100.0.4896.135

### Device architecture

arm64-v8a

### Android version

12

### Device model

lmi

### Changed flags

no flags changed

### Is this bug about the SystemWebView?

No

### Is this bug happening in an incognito tab?

Yes

### Is this bug caused by the adblocker?

No

### Is this bug a crash?

no

### Describe the bug

app_textures folder would not clean the jpeg when close the tab . 
i installed the chromium apk and tested , it did not have the problem. 
i think this bug may damage the privacy .

### Steps to reproduce the bug

when you open a tab and close it .

### Expected behavior

app_textures folder would clean .

**Comment by uazo** [comment]
when I close the tab the thumbail is deleted, if I close the app it doesn't, but when I restart it they are deleted.
can you give us the steps you use to reproduce it?

> Is this bug happening in an incognito tab?
> Yes

do you mean always incognito?

**Comment by inorinet** [comment]
if i click exit , i think it should clean all , but it still remain some jpeg now

**Comment by uazo** [comment]
> sometimes it truly clean it

it seems strange to me, you have to open the list of tabs to delete old files, or at least, the code (of chromium) foresees that. you can check?

> if i click exit , i think it should clean all , but it still remain some jpeg now

no, but it is normal, the exit in bromite is currently identical to the closure that is done through the swipe of the recents.

**Comment by inorinet** [comment]
i compare the chromium apk from https://download-chromium.appspot.com/ ,  when i close a tab , the behavior whether it delete  thumbail   is different.

**Comment by uazo** [comment]
> i compare the chromium apk from https://download-chromium.appspot.com/ 

that's the night build, please try with https://github.com/bromite/chromium/releases/tag/100.0.4896.92

**Comment by inorinet** [comment]
i checked it , it also has the problem, it seem like a upstream bug, but i am not very sure. i do not have the ability to check the source code about  chromium.

**Comment by csagan5** [comment]
@uazo is it reproducible with Chromium?

**Comment by uazo** [comment]
> is it reproducible with Chromium?

I don't see any bugs, what point do you mean?

**Comment by csagan5** [comment]
> > is it reproducible with Chromium?
> 
> I don't see any bugs, what point do you mean?

This one:
> sometimes it truly clean it , but not always, incognito mode has this bug more than normal mode

I suppose there should be no thumbnails anywhere saved to storage in incognito mode?

**Comment by uazo** [comment]
> I suppose there should be no thumbnails anywhere saved to storage in incognito mode?

the code currently provides for the generation of thumbnails also for the incognito mode, being used for the creation of the list of tabs during the selection.
I saw that there is also an enumeration that allows a tablist without any preview image, but I don't know if it is easy to activate it. among other things, I saw that the old tab list selection mode is still present.
I can verify it by looking at the kiwi code, because it seems to me that they have added a flag there to restore the old mode.

**Comment by csagan5** [comment]
> the code currently provides for the generation of thumbnails also for the incognito mode

And they would be saved to storage in incognito mode?

> among other things, I saw that the old tab list selection mode is still present.

URL?

> the old mode.

The old mode is ...?

**Comment by uazo** [comment]
> And they would be saved to storage in incognito mode?

if by memory you mean on disk, yes, they are saved

> URL?

https://source.chromium.org/chromium/chromium/src/+/main:chrome/android/features/tab_ui/java/src/org/chromium/chrome/browser/tasks/tab_management/TabListCoordinator.java;drc=b677ab23a658dd869ee77a2156a97e64bcb2f3ca;l=65

> The old mode is ...?

![image](https://user-images.githubusercontent.com/29201891/165959610-fc8f359e-a2cc-4b8d-aa41-fd35e9ad381b.png)

**Comment by uazo** [comment]
@csagan5 [here](https://uazo.github.io/superpatch/kiwi/commit/kiwi) you can see the code of kiwi above chromium, if you are interested

**Comment by csagan5** [comment]
> > And they would be saved to storage in incognito mode?
> 
> if by memory you mean on disk, yes, they are saved

Ok, so this is not great from privacy perspective but the bug is upstream, correct?

> a flag there to restore the old mode.

I guess we could also have a flag to use the old mode but how would it affect this bug here? It does not create thumbnails on storage?

**Comment by uazo** [comment]
> Ok, so this is not great from privacy perspective but the bug is upstream, correct?

correct.
if we activated the list mode we could deactivate the thumbnails, but they remain on the device...

> It does not create thumbnails on storage?

I don't think so, those are useful anyway, but I don't understand why they are not only in memory ..

**Comment by uazo** [comment]
> Both modes show a preview; what is the difference? Thumbnails are smaller and stored separately?

in fact the code saves two files per tab, but I don't know if they are in different quality

> Probably because of devices with little RAM. 

may be, it's always 350k per tab

> In the old mode are thumbnails created or not?

yes, it seems to be saved anyway

**Comment by csagan5** [comment]
Ok; we could address it but we should also ask upstream for the best course of action.

**Comment by csagan5** [comment]
> > URL?
> 
> https://source.chromium.org/chromium/chromium/src/+/main:chrome/android/features/tab_ui/java/src/org/chromium/chrome/browser/tasks/tab_management/TabListCoordinator.java;drc=b677ab23a658dd869ee77a2156a97e64bcb2f3ca;l=65
> 
> >

The line you highlighted is `GRID`; is that the old mode? I thought `GRID` is the new mode. There is a known upstream issue that the flag `#enable-grid-layout` cannot be disabled, so if it were possible to disable it we would get the old view back like this?

![image](https://user-images.githubusercontent.com/29201891/165959610-fc8f359e-a2cc-4b8d-aa41-fd35e9ad381b.png)

This feature restore is probably off-topic here and worth its own issue/PR if we want to re-add it; also, totally not related to privacy.

> I saw that there is also an enumeration that allows a tablist without any preview image, but I don't know if it is easy to activate it.

URL? I would like to do the privacy fix needed to close this issue (if there is any).

**Comment by uazo** [comment]
> if it were possible to disable it we would get the old view back like this?

it seems so, the value of the old mode is `CAROUSEL`

> also, totally not related to privacy.

that's right, I have other privacy patch in mind, I'd like to tell you about them

**Comment by csagan5** [comment]
> > I saw that there is also an enumeration that allows a tablist without any preview image, but I don't know if it is easy to activate it.
> 
> URL? I would like to do the privacy fix needed to close this issue (if there is any).

@uazo let me know how me/you can do the privacy fix so that this issue can also be closed

**Comment by csagan5** [comment]
@uazo ping about this issue, would still like to do a fix here.

**Comment by uazo** [comment]
>  would still like to do a fix here.

yes, you're right, you do well to remind me, but I haven't looked at it yet.
I was hoping to see how the kiwi guys did it, but I can't find their updated sources anymore (and among other things it seems to me that I have blocked my account in their repo, boh!)

**Comment by csagan5** [comment]
> yes, you're right, you do well to remind me, but I haven't looked at it yet.

What is the code saving the thumbnail? If you can point me in that direction maybe I can try to do it myself.

> it seems to me that I have blocked my account in their repo

Ignore that, some people need this to feel important.

**Comment by uazo** [comment]
>  If you can point me in that direction maybe I can try to do it myself.

Java side:
https://source.chromium.org/chromium/chromium/src/+/main:chrome/android/java/src/org/chromium/chrome/browser/compositor/layouts/content/TabContentManager.java;l=571;drc=aeedf4cd0d323c89781ecbe6662c899cdfa8aa81

Native side:
https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/android/compositor/tab_content_manager.cc;l=284;drc=aeedf4cd0d323c89781ecbe6662c899cdfa8aa81;bpv=1;bpt=1

Cache:
https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/thumbnail/cc/thumbnail_cache.cc;l=287;drc=aeedf4cd0d323c89781ecbe6662c899cdfa8aa81;bpv=1;bpt=1

it might also be interesting to understand what `LIST` does:
https://source.chromium.org/chromium/chromium/src/+/main:chrome/android/features/tab_ui/java/src/org/chromium/chrome/browser/tasks/tab_management/TabGridDialogCoordinator.java;l=82;drc=aeedf4cd0d323c89781ecbe6662c899cdfa8aa81;bpv=1;bpt=1
